kerchunk/grib2.py

```diff
@@ -582,3 +583,94 @@ def correct_hrrr_subhf_step(group: Dict) -> Dict:
     group["refs"]["step/0"] = enocded_val
 
     return group
+
+
+def parse_grib_idx(
+    fs: fsspec.AbstractFileSystem,
+    basename: str,
+    suffix: str = "idx",
+    tstamp: Optional[pd.Timestamp] = None,
+    validate: bool = False,
+) -> pd.DataFrame:
+    """
+    Standalone method used to extract metadata from a grib2 idx file(text) from NODD
+    :param fs: the file system to read from
+    :param basename: the base name is the full path to the grib file
+    :param suffix: the suffix is the ending for the idx file
+    :param tstamp: the timestamp to record for this index process
+    :return: the data frame containing the results
+    """
+
+    fname = f"{basename}.{suffix}"
+    fs.invalidate_cache(fname)
+    fs.invalidate_cache(basename)
+
+    baseinfo = fs.info(basename)
+
+    with fs.open(fname, "r") as f:
+        splits = []
+        for line in f.readlines():
+            try:
+                idx, offset, date, attrs = line.split(":", maxsplit=3)
+                splits.append(
+                    [
+                        int(idx),
+                        int(offset),
+                        f"{date[2:6]}-{date[6:8]}-{date[8:10]}-{date[10:]}",
+                        attrs,
+                    ]
+                )
+            except ValueError:
+                # Wrap the ValueError in a new one that includes the bad line
+                # If building the mapping, pick a different forecast run where the idx file is not broken
+                # If indexing a forecast using the mapping, fall back to reading the grib file
+                raise ValueError(f"Could not parse line: {line}")
+
+    result = pd.DataFrame(
+        data=splits,
+        columns=["idx", "offset", "date", "attrs"],
+    )
+
+    # Subtract the next offset to get the length using the filesize for the last value
+    result.loc[:, "length"] = (
```

Comment:
This and the following lines could be covered with a single call to pandas.DataFrame.assign